### ISSUE TITLE
Help reproducing pretrained checkpoint result: official checkpoint gives ~4.84% exact vs reported 15.26%

### ISSUE THREAD

**Issue by aboulaakoul-Elwalid** [ThanosM97/end-to-end-chess-recognition#5]
Hi Athanasios,
Thanks for releasing ChessReD and the official implementation. I am trying to reproduce the pretrained ResNeXt checkpoint result from the paper, but my local result is much lower than the reported number. I may be missing the exact evaluation script or checkpoint provenance.
## Expected result
The paper reports ResNeXt on the official ChessReD test split:
- boards with no mistakes: 15.26%
- mean incorrect squares: 3.40
- per-square error: 5.31%
## Checkpoint used
I downloaded the pretrained model linked in the README:
- Google Drive ID: `1sEkIj5MrFncGnmHQt66o_huKjqoMkNQ3`
- local SHA256: `a37eec7d804254b68aa317efef8e720f43113a62ab21380a9b4c8bc8d1d3bded`
- size: `1062664758` bytes
Checkpoint metadata includes:
- `epoch = 145`
- `global_step = 59130`
- `best_model_score / val_acc = 0.156021893`
- internal path: `best_model_epoch=145-val_acc=0.1560.ckpt`
## Dataset verification
The local dataset files match the official config:
- `annotations.json` MD5: `d34bca5ad46ec7a8df96a1d3c36784f3`
- `images.zip` / `chessred.zip` MD5: `32e23ed535239dc517a6499762d9847e`
## Public-code reproduction path
I ran the checkpoint using the closest public-code path I could reconstruct:
- public repo commit: `52f7a1922e388772c9658cb4edbe8542c1148deb`
- Python: `3.8.20`
- `torch==2.1.0`
- `torchvision==0.16.0`
- `pytorch-lightning==2.1.0`
- public `train.py`
- public `ChessResNeXt`
- `Trainer.predict(..., ckpt_path=...)`
- public `utils.recognition_accuracy`
## Local reproduction result
Using board-level exact match, i.e. 64/64 squares correct:
- val exact: `0.0483576655`
- test exact: `0.0483795218`
This is much lower than both:
- checkpoint metadata `val_acc ≈ 0.1560`
- paper test exact `0.1526`
## Checks already performed
I tried to rule out local evaluation mistakes:
- checkpoint loads strictly into the public architecture
- local model class and public repo class produce identical outputs:
  - `max_output_diff = 0.0`
- dataset loader parity against public `dataset.py` passes:
  - `max_abs_tensor_diff = 0.0`
  - `labels_equal = true`
- class order / square order / board rotations were checked:
  - identity `view(64, 13)` is best
## Question
Could you share one of the following?
1. The exact inference/evaluation script or command used to evaluate the pretrained checkpoint.
2. The expected SHA256/hash and filename of the official pretrained checkpoint.
3. The exact repo commit used to train/evaluate the released checkpoint.
4. Whether the released checkpoint was trained/evaluated with code that differs from the current public `train.py`.
5. Whether the checkpoint expects a different annotation version, split, class order, or label preprocessing.
One thing I noticed: the public `train.py` currently saves checkpoints with `monitor="val_loss"`, while the downloaded checkpoint metadata appears to come from a `val_acc`-monitored run (`best_model_epoch=145-val_acc=0.1560.ckpt`). Maybe I am missing the original evaluation/training path.
I have a compact reproduction bundle with the eval script, environment freeze, output summary, and first few predictions. Happy to share it if useful.
Thank you for any guidance.

**Comment by JSeytre** [comment]
Hi there! Thanks for this great work 👍 

I also encounter the same issue as @aboulaakoul-Elwalid and reproduce the exact same performance.
Please see the attached visualization which shows that **when the orientation of the chessboard is not from white's perspective, the outputs collapse**.

### Setup
I reproduced it two ways — my own wrapper, and your public `dataset.py` + `ChessResNeXt` + `recognition_accuracy` run directly — and got identical numbers.

Both give 4.84% exact / 29.8 mean incorrect squares on the official test split.

I investigated and found that the errors aren't uniform:
  - if they are white-side games (e.g. G000, G032, G088): ~2–4 incorrect squares (very similar to the paper's 3.40);
  - side-view or black-side games (e.g. G004, G018, G021): ~46–51 incorrect squares, and the prediction is garbage — see the attached screenshot

I also found that the outputs collapse on the training split too.
Running the checkpoint on one image per training game: 19/60 read perfectly (0 wrong), 41/60 collapse (~50 squares wrong). 

As a last verification, I started from your inference snippet in this repo's issue #2 : on a white-side start position our output exactly matches the expected tensor you posted there; on a black-side start position the same recipe returns 49/64 wrong. So quite possibly the checkpoint is wrong?

### Questions
These questions might help solve the issue:

  1. Was the released checkpoint trained on the current released ChessReD (our `annotations.json`/`images` match the official MD5s `d34b…` / `32e2…`), or an earlier version?
  2. Are we missing something in the pipeline? The failure is specifically when it's not viewed from white's side.
  3. The downloaded checkpoint's internal filename is `best_model_epoch=145-val_acc=0.1560.ckpt`. The `val_acc` in that name means it was saved by a training script that kept the best model by validation accuracy (and the number seems to match expectations of ~15% performance) — but the `train.py` in this repo keeps the best model by validation loss (`monitor="val_loss"`). So the released checkpoint might have been produced by a different script than the public one, can you verify this?

<img width="1210" height="2244" alt="Image" src="https://github.com/user-attachments/assets/82ae4433-3e73-4e25-a87d-7a5901d2a669" />

**Comment by tmasouris** [comment]
Hey @aboulaakoul-Elwalid and @JSeytre,

Sorry for the slow reply, @aboulaakoul-Elwalid, I appreciate your patience. And a big thank you to both of you for digging into this so carefully. The level of detail in your reports made it much easier for me to track down what went wrong.

You're both right. I reproduced the same ~4.84% myself. After going through old run logs and experiment files, I found the culprit: **an offline image preprocessing step that I forgot to document or include in the public release**. Apologies.

## What actually happened

To avoid running out of GPU memory during training, I pre-processed all images offline and resized them to **1024×1024** before training started. The internal training script therefore used a transform pipeline with **no resize**:

```python
transforms.Compose([
    transforms.ToPILImage(),
    transforms.ToTensor(),
    transforms.Normalize(mean=[...], std=[...]),
])
```

When I put the dataset together for public release, I decided to share the raw images so everyone would have access to the original data. To go with that, I added a `transforms.Resize(1024)` to the public `train.py`. That was a mistake. The checkpoint had never seen that transform, and it's not equivalent to what I did offline.

This is a textbook **distribution shift**. The normalization stats were computed on the offline 1024×1024 images, so feeding differently-sized inputs breaks that calibration from the start. The offline script also likely used a different interpolation kernel (e.g., Lanczos vs. torchvision's default bilinear), which affects the fine-grained texture cues the model relies on. All of this feeds into mismatched BatchNorm running statistics, and the error compounds across every layer in the network. 

## Reproducing the paper results

I wasn't able to find the original preprocessing script, but I did find the **pre-processed images** that were actually used for training and evaluation. You can grab them here: **[GDRIVE_LINK](https://drive.google.com/file/d/1jxmFxjOy0qefdCZ_x3DMNtsvAK4LojEw/view)**.

Using those images with the checkpoint and the transform **without** `Resize`, I get:

```
Board Recognition Accuracy (tolerance=0): 15.2%
Per-Square Accuracy: 94.59%
```

I've updated the README file to document this finding. Merging the PR will close this issue, but please reopen it if your are unable to reproduce the results using the preprocessed images.

## On the `val_acc` vs `val_loss` checkpoint naming

The actual training script (model architecture, optimizer, loss function, data pipeline) was identical to the one on this repository. The only differences were in the checkpoint callback config (which metric was used in saving) and some logging details.

## What's next

I'm going to reach out to TU Delft to get the pre-processed images added to the official dataset release so this isn't a problem for anyone else. In the meantime, the link above should get you going.

Thanks again for pointing me to this issue. I hope the work is useful for whatever you're building or writing. I'd love to hear about it!

**Comment by joelkartoon** [comment]
Using the updated data, I do get performance numbers similar to what you report. Thanks!

I am working on a model that is training / testing on your data. I'll make sure to let you know when I am ready to share information about it :)

Thanks for your help, and for sharing this dataset & model with the community!
